### PR TITLE
[AutoDiff] Introduce `autodiff_function` instruction

### PR DESCRIPTION
**Pull request by rxwei**
### Introduction

An `autodiff_function` instruction bundles a function operand with its associated differentiation functions up to a specified differentiation order into an ``@autodiff`` function. Normally, there are 4 associated functions per differentiation order: a Jacobian-vector products (JVP) function, a thick differential function that acts as the JVP function's result, vector-Jacobian products (VJP) function, and a thick pullback function that acts as the VJP function's result. When ``[legacy_reverse]`` is specified, only first-order reverse-mode differentiation is legal, and only two functions are being bundled: the primal function and the adjoint function.

This instruction will be triggered by an `AutoDiffFunctionExpr` (not yet introduced) that converts a non-differentiable function to an `@autodiff` function. Here is an example of when that happens:
```swift
func bar(_: Float) -> Float { ... }

func diff(_ f: @autodiff (Float) -> Float) {
  _ = gradient(of: f)
}

diff(bar) // (apply_expr (autodiff_function_expr "diff" args=("bar")))
```

### Syntax
```bnf
sil-instruction ::= 'autodiff_function'
                    sil-autodiff-function-legacy-reverse-mode?
                    sil-autodiff-function-parameter-indices?
                    sil-autodiff-function-order?
                    sil-value ':' sil-type
                    sil-autodiff-associated-functions-clause?
                      
sil-autodiff-function-legacy-reverse-mode ::= '[' 'legacy_reverse' ']'
sil-autodiff-function-parameter-indices ::= '[' 'wrt' [0-9]+ (',', [0-9]+)* ']'
sil-autodiff-function-order ::= '[' 'order' [0-9]+ ']'
sil-autodiff-associated-functions-clause ::= 'with' sil-autodiff-associated-function-list
                                             (',' sil-autodiff-associated-function-list)*
sil-autodiff-associated-function-list ::= '{' sil-value (',' sil-value)* '}'
```

### Examples
```swift
  autodiff_function [wrt 0] [order 1] %0 : $(T) -> T
      with {%1 : $(T) -> (T) -> T, %2 : $(T, @box {...}) -> T, %3 : $(T) -> (T) -> T, %4 : $(T, @box {...}) -> T}

  autodiff_function [legacy_reverse] [wrt 0] [order 1] %0 : $(T) -> T
      with {%1 : $(T) -> (..., T), %2 : $(T, ..., T, T) -> T}
```

### Semantics

``[parameters ...]`` specifies parameter indices that the original function is differentiable with respect to. When not specified, it defaults to all parameters.

``[order ...]`` specifies the maximum differentiation order for the resulting function. The number of lists of associated functions is equal to the order.

A ``with`` clause specifies the differentiation functions associated with the original function. When a ``with`` clause is not specified, the first operand will be differentiated to produce associated functions, and a ``with`` clause will be added to the instruction.

### Bundle Layout

The layout of an `@autodiff` function starts with the original function, followed by parameter indices, the differentiation order, and a flat list of associated functions.

1. Standard mode AD. The implementation of the future design described in [the manifesto](https://gist.github.com/rxwei/30ba75ce092ab3b0dce4bde1fc2c9f1d) will use this semantics. 
    ```
                   Order 1                  Order 2                 ...
    |----------| |-----|----|-----|----| |-----|----|-----|----| ...
    | Original | | JVP | DF | VJP | PB | | JVP | DF | VJP | PB | ...
    |----------| |-----|----|-----|----| |-----|----|-----|----| ...
    ```
    Note: "DF" means differential. "PB" means pullback.

2. Legacy (existing) reverse mode AD. This is to support today's reverse-mode AD and will be deprecated when the standard mode is implemented.
    ```
                   Order 1
    |----------| |--------|---------|
    | Original | | Primal | Adjoint |
    |----------| |--------|---------|
    ```

### CODE REVIEW

File: include/swift/SIL/SILCloner.h [hunk 1]
```diff
@@ -873,6 +873,27 @@ SILCloner<ImplClass>::visitGradientInst(GradientInst *Inst) {
...
 }
 
+template<typename ImplClass>
+void
+SILCloner<ImplClass>::visitAutoDiffFunctionInst(AutoDiffFunctionInst *Inst) {
+  getBuilder().setCurrentDebugScope(getOpScope(Inst->getDebugScope()));
+//  SILValue originalFunction,
+//  const SmallBitVector &parameterIndices,
+//  bool isLegacyReverseMode, unsigned differentiationOrder,
+//  ArrayRef<SILValue> associatedFunctions);
```
Comment (dan-zheng) [inline include/swift/SIL/SILCloner.h]: Remove comments?

**Review by dan-zheng — APPROVED** [review]
Autodiff semantics and implementation LGTM!

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**